### PR TITLE
fix(graphql): modify fields match logic for hasOne directive when using auth directive

### LINKED ISSUE #9401: `@hasOne` field causes push operations to fail, when the referenced type has a`@primaryKey` containing 2 (or more) sort key fields.
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-cli/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### How did you install the Amplify CLI?

npm

### If applicable, what version of Node.js are you using?

v17.2.0

### Amplify CLI Version

7.6.5

### What operating system are you using?

macOS 12.0.1

### Amplify Categories

api

### Amplify Commands

push

### Describe the bug

I am working on migrating my app to the v2 amplify GraphQL transformer. My setup was working fine with the v1 transformer.
When the `@hasOne` field described in this bug is commented out, the project "compiles" and pushes correctly.

One of my types has a primary key, with 2 sort key fields; see attached partial schema.

When running
`amplify api push --yes` or `amplify push --yes`, I get the following error:
```
✖ An error occurred when pushing the resources to the cloud
🛑 Invalid @hasOne on Post:associatedTawk. Provided fields do not match the size of primary key(s) for TawkCandidate
InvalidDirectiveError: Invalid @hasOne on Post:associatedTawk. Provided fields do not match the size of primary key(s) for TawkCandidate
    at getRelationalPrimaryMap (/opt/homebrew/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/graphql-auth-transformer/src/utils/schema.ts:129:13)
    at AuthTransformer.protectRelationalResolver (/opt/homebrew/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/graphql-auth-transformer/src/graphql-auth-transformer.ts:526:59)
    at AuthTransformer.generateResolvers (/opt/homebrew/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/graphql-auth-transformer/src/graphql-auth-transformer.ts:348:16)
    at GraphQLTransform.transform (/opt/homebrew/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/graphql-transformer-core/src/transformation/transform.ts:252:21)
    at _buildProject (/opt/homebrew/lib/node_modules/@aws-amplify/cli/node_modules/amplify-provider-awscloudformation/src/graphql-transformer/transform-graphql-schema.ts:563:37)
    at buildAPIProject (/opt/homebrew/lib/node_modules/@aws-amplify/cli/node_modules/amplify-provider-awscloudformation/src/graphql-transformer/transform-graphql-schema.ts:520:24)
    at transformGraphQLSchema (/opt/homebrew/lib/node_modules/@aws-amplify/cli/node_modules/amplify-provider-awscloudformation/src/graphql-transformer/transform-graphql-schema.ts:393:29)
    at Object.run (/opt/homebrew/lib/node_modules/@aws-amplify/cli/node_modules/amplify-provider-awscloudformation/src/push-resources.ts:166:5)
    at async Promise.all (index 0)
    at providersPush (/opt/homebrew/lib/node_modules/@aws-amplify/cli/src/extensions/amplify-helpers/push-resources.ts:139:3)
    at AmplifyToolkit.pushResources [as _pushResources] (/opt/homebrew/lib/node_modules/@aws-amplify/cli/src/extensions/amplify-helpers/push-resources.ts:96:9)
    at Object.executeAmplifyCommand (/opt/homebrew/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/amplify-category-api/src/index.ts:223:5)
    at executePluginModuleCommand (/opt/homebrew/lib/node_modules/@aws-amplify/cli/src/execution-manager.ts:178:3)
    at executeCommand (/opt/homebrew/lib/node_modules/@aws-amplify/cli/src/execution-manager.ts:30:5)
    at Object.run (/opt/homebrew/lib/node_modules/@aws-amplify/cli/src/index.ts:205:5)
```

After some digging around with a local clone of `amplify-cli`, it looks like in the vicinity of
`@aws-amplify/graphql-auth-transformer/src/utils/schema.ts:129:13` (the first line in the stack trace above), 
the variable `const relatedPrimaryFields` (defined on line 125) contains the following payload:
`ownerSub,localId#tawkVersion`

I.e., `TawkCandidate`'s 2 sort keys (`localId` and `tawkVersion`) occupy a **single** element in the array joined with a `"#"` character -- throwing off the argument length comparison and triggering an error.

### Expected behavior

The app should "compile" and push successfully...

### Reproduction steps

Run `amplify api push --yes` or `amplify push --yes`

### GraphQL schema(s)

<details>

```graphql
# Put schemas below this line

type TawkCandidate
  @model(subscriptions: {level: off})
  @auth(
     rules: [
       {allow: owner, ownerField: "ownerSub", identityClaim: "sub", operations: [read]},
       {allow: public, operations: [read]},
       {allow: private, provider: iam, operations: [create, read, update, delete]}
     ]
) {
  ownerSub: String!
    @primaryKey(sortKeyFields: ["localId", "tawkVersion"])
    @index(name: "byAuthor", sortKeyFields: ["createdAt"])
  localId: String!
  tawkVersion: String!
  # ...
}

# ...

type Post
  @model(subscriptions: {level: off})
  @searchable
  @auth(
     rules: [
      {allow: owner, ownerField: "ownerSub", identityClaim: "sub", operations: [read, update, delete]},
      {allow: public, operations: [read]},
      {allow: private, provider: iam, operations: [create, read, update, delete]}
    ]
) {
  id: ID!
  ownerSub: String! @index(name: "byAuthor", sortKeyFields: ["createdAt"])

  associatedTawk_localId: String
  associatedTawk_tawkVersion: String
  associatedTawk: TawkCandidate @hasOne(fields: ["ownerSub", "associatedTawk_localId", "associatedTawk_tawkVersion"])
  # ...
}


```

</details>

### Log output

<details>

```
# Put your logs below this line

If this would be of assistance, please advise which kinds of "sensitive data" may appear in the logs...
At a glance I don't see any but I don't want to accidentally miss some.
```

</details>

### Additional information

As noted above:

After some digging around with a local clone of `amplify-cli`, it looks like in the vicinity of
`@aws-amplify/graphql-auth-transformer/src/utils/schema.ts:129:13` (the first line in the stack trace above), 
the variable `const relatedPrimaryFields` (defined on line 125) contains the following payload:
`ownerSub,localId#tawkVersion`

I.e., `TawkCandidate`'s 2 sort keys (`localId` and `tawkVersion`) occupy a **single** element in the array joined with a `"#"` character -- throwing off the argument length comparison and triggering an error.

### PR DESCRIPTION
**Pull request by sundersc**
#### Description of changes
Using hasOne directive with @auth throws an error for primary keys with multiple sort keys. This PR will fix the issue https://github.com/aws-amplify/amplify-cli/issues/9401.

#### Issue #, if available
https://github.com/aws-amplify/amplify-cli/issues/9401

#### Description of how you validated changes
- yarn test
- manual test

#### Checklist
- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-graphql-auth-transformer/src/__tests__/relational-auth.test.ts [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+      type Comment
+        @model
+        @auth(rules: [
+          {allow: owner},
+          {allow: groups, groups: ["Moderator"]}
+      ]) {
+        id: ID!
+        text: String
+        post: Post @belongsTo
+      }`;
```
[the remark this answers is not in the document]
Comment (sundersc) [reply packages/amplify-graphql-auth-transformer/src/__tests__/relational-auth.test.ts:32]: updated

File: packages/amplify-graphql-auth-transformer/src/__tests__/relational-auth.test.ts [hunk 2]
```diff
@@ -0,0 +1,248 @@
+import { AuthTransformer } from '../graphql-auth-transformer';
+import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
+import { HasOneTransformer, HasManyTransformer, BelongsToTransformer } from '@aws-amplify/graphql-relational-transformer'
```
[the remark this answers is not in the document]
Comment (sundersc) [reply packages/amplify-graphql-auth-transformer/src/__tests__/relational-auth.test.ts:3]: Moved the test to `@aws-amplify/graphql-relational-transformer` package.

File: packages/amplify-graphql-auth-transformer/src/__tests__/relational-auth.test.ts [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+      new ModelTransformer(),
+      new HasManyTransformer(),
+      new BelongsToTransformer(),
+      new AuthTransformer(),
+    ],
+  });
+  let out;
+  expect(() => {
+    out = transformer.transform(validSchema);
+  }).not.toThrowError();
```
[the remark this answers is not in the document]
Comment (sundersc) [reply packages/amplify-graphql-auth-transformer/src/__tests__/relational-auth.test.ts:51]: updated

File: packages/amplify-graphql-auth-transformer/src/__tests__/relational-auth.test.ts [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+      new HasOneTransformer(),
+      new HasManyTransformer(),
+      new BelongsToTransformer(),
+      new AuthTransformer(),
+    ],
+  });
+  let out;
+  expect(() => {
+    out = transformer.transform(validSchema);
+  }).not.toThrowError();
```
[the remark this answers is not in the document]
Comment (sundersc) [reply packages/amplify-graphql-auth-transformer/src/__tests__/relational-auth.test.ts:149]: updated

File: packages/amplify-graphql-auth-transformer/src/__tests__/relational-auth.test.ts [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
```
[the remark this answers is not in the document]
Comment (sundersc) [reply packages/amplify-graphql-auth-transformer/src/__tests__/relational-auth.test.ts:247]: updated

File: packages/amplify-graphql-relational-transformer/src/__tests__/relational-auth.test.ts [hunk 6]
```diff
@@ -180,3 +181,235 @@ const expectTwo = (fieldOrType: any, directiveNames: string[]) => {
...
 };
+
+test('auth with hasMany relation - only partition key', () => {
+  const validSchema = `
+      type Post
+        @model
+        @auth(rules: [
+                {allow: owner},
+                {allow: groups, groups: ["Moderator"]}
+            ]) {
```
Comment (cjihrig) [inline packages/amplify-graphql-relational-transformer/src/__tests__/relational-auth.test.ts:192]: Indentation is off on these 3 lines.

File: packages/amplify-graphql-relational-transformer/src/__tests__/relational-auth.test.ts [hunk 7]
```diff
@@ -180,3 +181,235 @@ const expectTwo = (fieldOrType: any, directiveNames: string[]) => {
...
+  const modelPostConnectionType = getObjectType(schemaDoc, 'ModelPostConnection');
+  expect(modelPostConnectionType).toBeDefined();
+});
+
+test('auth with hasOne relation mismatch fields count - missing sort key must throw an error', () => {
+  const validSchema = `
+      type Student
+        @model
+        @auth(rules: [
+            {allow: owner}
```
Comment (cjihrig) [inline packages/amplify-graphql-relational-transformer/src/__tests__/relational-auth.test.ts:232]: Indentation is off here.

File: packages/amplify-graphql-relational-transformer/src/__tests__/relational-auth.test.ts [hunk 8]
```diff
@@ -180,3 +181,235 @@ const expectTwo = (fieldOrType: any, directiveNames: string[]) => {
...
+    out = transformer.transform(validSchema);
+  }).toThrowError('Invalid @hasOne on Student:scores. Provided fields do not match the size of primary key(s) for StudentScore');
+});
+
+test('auth with hasOne relation match fields count - single sort key do not throw error', () => {
+  const validSchema = `
+    type Student
+      @model
+      @auth(rules: [
+          {allow: owner}
```
Comment (cjihrig) [inline packages/amplify-graphql-relational-transformer/src/__tests__/relational-auth.test.ts:279]: Same here.

File: packages/amplify-graphql-relational-transformer/src/__tests__/relational-auth.test.ts [hunk 9]
```diff
@@ -180,3 +181,235 @@ const expectTwo = (fieldOrType: any, directiveNames: string[]) => {
...
+  // Graphql transform should not throw an error
+  const out = transformer.transform(validSchema);
+});
+
+test('auth with hasOne relation mismatch fields count - partial missing sort key must throw an error', () => {
+  const validSchema = `
+    type Student
+      @model
+      @auth(rules: [
+          {allow: owner}
```
Comment (cjihrig) [inline packages/amplify-graphql-relational-transformer/src/__tests__/relational-auth.test.ts:325]: And here.

File: packages/amplify-graphql-relational-transformer/src/__tests__/relational-auth.test.ts [hunk 10]
```diff
@@ -180,3 +181,235 @@ const expectTwo = (fieldOrType: any, directiveNames: string[]) => {
...
+    out = transformer.transform(validSchema);
+  }).toThrowError('Invalid @hasOne directive on scores. Partial sort keys are not accepted.');
+});
+
+test('auth with hasOne relation match fields count - multiple sort keys do not throw error', () => {
+  const validSchema = `
+    type Student
+      @model
+      @auth(rules: [
+          {allow: owner}
```
Comment (cjihrig) [inline packages/amplify-graphql-relational-transformer/src/__tests__/relational-auth.test.ts:374]: And here.

**Review by cjihrig — APPROVED** [review]
Still some minor style issues, but LGTM

**Comment by ataibarkai** [comment]
@sundersc I just submitted a new bug report (https://github.com/aws-amplify/amplify-cli/issues/9573) which manifests with a `hasMany` field -- the offending line is the one preceding the line fixed in this issue.

I wonder if a similar fixed is required there?

**Pull request merged**